File: backend/app/domain.py
from enum import Enum


class ShipmentStatus(str, Enum):
    CREATED = "created"
    # Origin leg
    PICKED_UP = "picked_up"
    CONSOLIDATED = "consolidated"
    DRAYAGE = "drayage"
    EXPORT_CUSTOMS = "export_customs"
    # Main haul
    HANDOVER = "handover"
    IN_TRANSIT = "in_transit"
    # Destination leg
    IMPORT_CUSTOMS = "import_customs"
    INLAND_HAUL = "inland_haul"
    DELIVERED = "delivered"
    CLOSED_OUT = "closed_out"


class CustomsStatus(str, Enum):
    DECLARATION_FILED = "declaration_filed"
    UNDER_REVIEW = "under_review"
    HELD_QUERY = "held_query"
    RELEASED = "released"
    ESCALATED = "escalated"


class EventType(str, Enum):
    # Origin leg
    CARGO_PICKED_UP = "cargo_picked_up"
    CONSOLIDATED = "consolidated"
    AT_PORT = "at_port"
    # Customs
    EXPORT_DECLARATION_FILED = "export_declaration_filed"
    IMPORT_DECLARATION_FILED = "import_declaration_filed"
    CUSTOMS_RELEASED = "customs_released"
    CUSTOMS_HELD = "customs_held"
    QUERY_RESPONDED = "query_responded"
    # Main haul
    HANDOVER = "handover"
    TRANSIT_PING = "transit_ping"
    ARRIVED = "arrived"
    # Destination leg
    INLAND_DISPATCHED = "inland_dispatched"
    DELIVERED = "delivered"
    CLOSED_OUT = "closed_out"


# Events that advance the main milestone chain
MILESTONE_EVENTS: set[EventType] = {
    EventType.CARGO_PICKED_UP,
    EventType.CONSOLIDATED,
    EventType.AT_PORT,
    EventType.HANDOVER,
    EventType.TRANSIT_PING,
    EventType.ARRIVED,
    EventType.INLAND_DISPATCHED,
    EventType.DELIVERED,
    EventType.CLOSED_OUT,
}

# Events that belong to the customs child workflow
CUSTOMS_EVENTS: set[EventType] = {
    EventType.EXPORT_DECLARATION_FILED,
    EventType.IMPORT_DECLARATION_FILED,
    EventType.CUSTOMS_RELEASED,
    EventType.CUSTOMS_HELD,
    EventType.QUERY_RESPONDED,
}

# Which events are legal given the current shipment status
ALLOWED_EVENTS: dict[ShipmentStatus, list[EventType]] = {
    ShipmentStatus.CREATED: [EventType.CARGO_PICKED_UP],
    ShipmentStatus.PICKED_UP: [EventType.CONSOLIDATED],
    ShipmentStatus.CONSOLIDATED: [EventType.AT_PORT],
    ShipmentStatus.DRAYAGE: [EventType.EXPORT_DECLARATION_FILED],
    ShipmentStatus.EXPORT_CUSTOMS: [
        EventType.EXPORT_DECLARATION_FILED,  # opens the gate (child's first recv)
        EventType.CUSTOMS_RELEASED,
        EventType.CUSTOMS_HELD,
        EventType.QUERY_RESPONDED,
    ],
    ShipmentStatus.HANDOVER: [EventType.TRANSIT_PING, EventType.ARRIVED],
    ShipmentStatus.IN_TRANSIT: [EventType.TRANSIT_PING, EventType.ARRIVED],
    ShipmentStatus.IMPORT_CUSTOMS: [
        EventType.IMPORT_DECLARATION_FILED,
        EventType.CUSTOMS_RELEASED,
        EventType.CUSTOMS_HELD,
        EventType.QUERY_RESPONDED,
    ],
    ShipmentStatus.INLAND_HAUL: [EventType.DELIVERED],
    ShipmentStatus.DELIVERED: [EventType.CLOSED_OUT],
    ShipmentStatus.CLOSED_OUT: [],
}

# Human-readable labels for the UI
STATUS_LABELS: dict[ShipmentStatus, str] = {
    ShipmentStatus.CREATED: "Created",
    ShipmentStatus.PICKED_UP: "Picked Up",
    ShipmentStatus.CONSOLIDATED: "Consolidated",
    ShipmentStatus.DRAYAGE: "Drayage to Port",
    ShipmentStatus.EXPORT_CUSTOMS: "Export Customs",
    ShipmentStatus.HANDOVER: "Carrier Handover",
    ShipmentStatus.IN_TRANSIT: "In Transit",
    ShipmentStatus.IMPORT_CUSTOMS: "Import Customs",
    ShipmentStatus.INLAND_HAUL: "Inland Haul",
    ShipmentStatus.DELIVERED: "Delivered",
    ShipmentStatus.CLOSED_OUT: "Closed Out",
}

EVENT_LABELS: dict[EventType, str] = {
    EventType.CARGO_PICKED_UP: "Cargo Picked Up",
    EventType.CONSOLIDATED: "Consolidated at Warehouse",
    EventType.AT_PORT: "Arrived at Port (Drayage Complete)",
    EventType.EXPORT_DECLARATION_FILED: "Export Declaration Filed",
    EventType.IMPORT_DECLARATION_FILED: "Import Declaration Filed",
    EventType.CUSTOMS_RELEASED: "Customs Released",
    EventType.CUSTOMS_HELD: "Customs Hold Issued",
    EventType.QUERY_RESPONDED: "Query Response Submitted",
    EventType.HANDOVER: "Handed Over to Carrier",
    EventType.TRANSIT_PING: "In-Transit Ping",
    EventType.ARRIVED: "Arrived at Destination Port",
    EventType.INLAND_DISPATCHED: "Dispatched Inland",
    EventType.DELIVERED: "Delivered (PoD Captured)",
    EventType.CLOSED_OUT: "Shipment Closed Out",
}

# Customs gate statuses (when a status is a gate, we run a child customs workflow)
CUSTOMS_GATE_STATUSES: set[ShipmentStatus] = {
    ShipmentStatus.EXPORT_CUSTOMS,
    ShipmentStatus.IMPORT_CUSTOMS,
}

MAX_CUSTOMS_ATTEMPTS = 3

File: frontend/src/domain.ts
import type { EventType, ShipmentStatus } from "./types";

export const STATUS_LABELS: Record<ShipmentStatus, string> = {
  created: "Created",
  picked_up: "Picked Up",
  consolidated: "Consolidated",
  drayage: "Drayage to Port",
  export_customs: "Export Customs",
  handover: "Carrier Handover",
  in_transit: "In Transit",
  import_customs: "Import Customs",
  inland_haul: "Inland Haul",
  delivered: "Delivered",
  closed_out: "Closed Out",
};

export const EVENT_LABELS: Record<EventType, string> = {
  cargo_picked_up: "Cargo Picked Up",
  consolidated: "Consolidated at Warehouse",
  at_port: "Arrived at Port (Drayage Complete)",
  export_declaration_filed: "Export Declaration Filed",
  import_declaration_filed: "Import Declaration Filed",
  customs_released: "Customs Released",
  customs_held: "Customs Hold Issued",
  query_responded: "Query Response Submitted",
  handover: "Handed Over to Carrier",
  transit_ping: "In-Transit Ping",
  arrived: "Arrived at Destination Port",
  inland_dispatched: "Dispatched Inland",
  delivered: "Delivered (PoD Captured)",
  closed_out: "Shipment Closed Out",
};

export const ALLOWED_EVENTS: Record<ShipmentStatus, EventType[]> = {
  created: ["cargo_picked_up"],
  picked_up: ["consolidated"],
  consolidated: ["at_port"],
  drayage: ["export_declaration_filed"],
  export_customs: [
    "export_declaration_filed",
    "customs_released",
    "customs_held",
    "query_responded",
  ],
  handover: ["transit_ping", "arrived"],
  in_transit: ["transit_ping", "arrived"],
  import_customs: [
    "import_declaration_filed",
    "customs_released",
    "customs_held",
    "query_responded",
  ],
  inland_haul: ["delivered"],
  delivered: ["closed_out"],
  closed_out: [],
};

// The ordered list of statuses in the chain (excluding "created")
export const STATUS_ORDER: ShipmentStatus[] = [
  "picked_up",
  "consolidated",
  "drayage",
  "export_customs",
  "handover",
  "in_transit",
  "import_customs",
  "inland_haul",
  "delivered",
  "closed_out",
];

export const ORIGIN_LEG: ShipmentStatus[] = [
  "picked_up",
  "consolidated",
  "drayage",
  "export_customs",
];
export const HAUL_LEG: ShipmentStatus[] = ["handover", "in_transit"];
export const DEST_LEG: ShipmentStatus[] = [
  "import_customs",
  "inland_haul",
  "delivered",
  "closed_out",
];

export const CUSTOMS_STATUSES = new Set<ShipmentStatus>([
  "export_customs",
  "import_customs",
]);

export type NodeState = "completed" | "active" | "customs" | "pending";

export function getNodeState(
  nodeStatus: ShipmentStatus,
  currentStatus: ShipmentStatus
): NodeState {
  const nodeIdx = STATUS_ORDER.indexOf(nodeStatus);
  const curIdx = STATUS_ORDER.indexOf(currentStatus);

  if (nodeIdx < curIdx) return "completed";
  if (nodeIdx === curIdx) {
    return CUSTOMS_STATUSES.has(nodeStatus) ? "customs" : "active";
  }
  return "pending";
}

// Colors matching the SVG palette
export const NODE_COLORS: Record<NodeState, { bg: string; border: string; text: string; sub: string }> = {
  completed: { bg: "#e1f5ee", border: "#0f6e56", text: "#085041", sub: "#0f6e56" },
  active:    { bg: "#e6f1fb", border: "#185fa5", text: "#0c447c", sub: "#185fa5" },
  customs:   { bg: "#faeedа", border: "#854f0b", text: "#633806", sub: "#854f0b" },
  pending:   { bg: "#f1efe8", border: "#c5c4bc", text: "#5f5e5a", sub: "#9f9e98" },
};

// Amber doesn't render so fallback
export const NODE_COLORS_SAFE: Record<NodeState, { bg: string; border: string; text: string; sub: string }> = {
  completed: { bg: "#e1f5ee", border: "#0f6e56", text: "#085041", sub: "#0f6e56" },
  active:    { bg: "#e6f1fb", border: "#185fa5", text: "#0c447c", sub: "#185fa5" },
  customs:   { bg: "#fff3cd", border: "#854f0b", text: "#633806", sub: "#854f0b" },
  pending:   { bg: "#f1efe8", border: "#c5c4bc", text: "#5f5e5a", sub: "#9f9e98" },
};

export const STATUS_SUBTITLES: Partial<Record<ShipmentStatus, string>> = {
  picked_up: "first mile",
  consolidated: "store, pack",
  drayage: "to port",
  export_customs: "customs gate",
  handover: "B/L or AWB",
  in_transit: "en route",
  import_customs: "customs gate",
  inland_haul: "to receiver",
  delivered: "PoD captured",
  closed_out: "reconcile",
};
